File: user-app/src/main/java/com/github/dkurata38/liblio/user_app/security/LiblioUserDetailsService.java
package com.github.dkurata38.liblio.user_app.security;

import lombok.RequiredArgsConstructor;

import java.util.List;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import com.github.dkurata38.liblio.domain.user.UserRepository;
import com.github.dkurata38.liblio.domain.user.Username;

@RequiredArgsConstructor
public class LiblioUserDetailsService implements UserDetailsService {
    private final UserRepository userRepository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return userRepository.findByUsername(new Username(username))
            .map(user -> new LiblioUser(
                user.getUserId().getValue(),
                user.getUsername().getValue(),
                user.getPassword().getValue(),
                List.of(new SimpleGrantedAuthority("ROLE_USER"))
            ))
            .orElseThrow(() -> new UsernameNotFoundException("User is not found"));
    }
}

File: user-app/src/main/kotlin/com/github/dkurata38/liblio/user_app/security/LiblioUserDetailsService.kt
package com.github.dkurata38.liblio.user_app.security

import com.github.dkurata38.liblio.domain.user.UserRepository
import com.github.dkurata38.liblio.domain.user.Username
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException

class LiblioUserDetailsService(private val userRepository: UserRepository): UserDetailsService {
    override fun loadUserByUsername(username: String?): UserDetails {
        if (username == null) {
            throw NullPointerException()
        }
        val user = userRepository.findByUsername(Username(username)) ?: throw UsernameNotFoundException("User is not found")
        return LiblioUser(
            user.userId.value,
            user.username.value,
            user.password.value,
            mutableListOf(SimpleGrantedAuthority("ROLE_USER"))
        )
    }
}
